### ISSUE TITLE
vmq bridge module crashed and did not restart

### ISSUE THREAD

**Issue by mravkine** [erlio/vernemq#785]
### Environment

- VerneMQ Version: 1.4.1
- OS: Ubuntu 16.04.4 LTS
- Erlang/OTP version (if building from source): docker image

### Expected behavior

I am using VMQ to bridge to IBM Watson IoT. My configuration is as follows:

```
vmq_bridge.tcp.br0 = surmov.messaging.internetofthings.ibmcloud.com:1883
vmq_bridge.tcp.br0.client_id = a:surmov:mosquitto-nucleus1
vmq_bridge.tcp.br0.username = a-surmov-...
vmq_bridge.tcp.br0.password = ....
vmq_bridge.tcp.br0.try_private = off
vmq_bridge.tcp.br0.cleansession = on

vmq_bridge.tcp.br0.keepalive_interval = 10
vmq_bridge.tcp.br0.restart_timeout = 5

vmq_bridge.tcp.br0.topic.1 = iot-2/type/+/id/+/evt/+/fmt/json in
vmq_bridge.tcp.br0.topic.2 = iot-2/type/+/id/+/mon in
```
It was working fine for 2 weeks, delivering over 600M messages.

### Actual behaviour

Last night at 3am, this happened:

```
07:54:35.270 [error] gen_fsm <0.413.0> in state waiting_for_connack terminated with reason: no function clause matching gen_emqtt:handle_frame(waiting_for_connack, {mqtt_publish,unde
fined,[<<"iot-2">>,<<"type">>,<<"atheros-csi">>,<<"id">>,<<"csi-g-883d249d6fee">>,...],...}, {state,"surmov.messaging.internetofthings.ibmcloud.com",1883,#Port<0.459156>,101,"a-surmo
v-zie2b...",...}) line 368
07:54:35.270 [error] CRASH REPORT Process <0.413.0> with 2 neighbours exited with reason: no function clause matching gen_emqtt:handle_frame(waiting_for_connack, {mqtt_publish,undefi
ned,[<<"iot-2">>,<<"type">>,<<"atheros-csi">>,<<"id">>,<<"csi-g-883d249d6fee">>,...],...}, {state,"surmov.messaging.internetofthings.ibmcloud.com",1883,#Port<0.459156>,101,"a-surmov-
zie2b...",...}) line 368 in gen_fsm:terminate/7 line 559
07:54:35.271 [error] Supervisor vmq_bridge_sup had child {vmq_bridge,"surmov.messaging.internetofthings.ibmcloud.com",1883} started with vmq_bridge:start_link("surmov.messaging.inter
netofthings.ibmcloud.com", 1883, {vmq_reg,direct_plugin_exports,[vmq_bridge]}) at <0.410.0> exit with reason no function clause matching gen_emqtt:handle_frame(waiting_for_connack, {
mqtt_publish,undefined,[<<"iot-2">>,<<"type">>,<<"atheros-csi">>,<<"id">>,<<"csi-g-883d249d6fee">>,...],...}, {state,"surmov.messaging.internetofthings.ibmcloud.com",1883,#Port<0.459
156>,101,"a-surmov-zie2b...",...}) line 368 in context child_terminated
07:54:35.272 [debug] transition from offline --> online because of add_session
07:54:35.272 [debug] Supervisor vmq_bridge_sup started vmq_bridge:start_link("surmov.messaging.internetofthings.ibmcloud.com", 1883, {vmq_reg,direct_plugin_exports,[vmq_bridge]}) at 
pid <0.770.25>
```

The worst part:  It says the bridge was restarted but it was NOT actually restarted .. I had to manually restart the VMQ docker this morning to make it reconnect to the bridge.

**Comment by ioolkos** [comment]
@mravkine thanks for opening the issue, Mike!
@dergraf @larshesel can you give a hand in analysing this? anything obvious from that state machine terminate?

**Comment by larshesel** [comment]
Hi @mravkine The bridge finite state machine errors out because it's receiving an MQTT publish message while it's doing the connection handshake (it's waiting for the connack from the broker). This is explicitly forbidden by the MQTT 3.1.1 spec: `The first packet sent from the Server to the Client MUST be a CONNACK Packet [MQTT-3.2.0-1].` so I don't think we'll try to handle that.

I'll see if I can verify/fix the restart issue.

**Comment by larshesel** [comment]
@dergraf I think I spotted the problem - when the bridge client is restarted there's nothing calling the `vmq_bridge:setopts` function. This function is called when the bridge is correctly started and if the bridge config changes - and the idea is that if the changes only consist of changes to subscriptions then the connection is maintained. However, I don't believe we actually expose this functionality (changing the bridge topics) anywhere.

I'm therefore preparing a PR where we remove the `setopts` function and make the options part of the arguments to the child process so it can be properly restarted by the supervisor. If then the options change then we simply remove the child process and and add it again with the new options (and hence reconnecting in all cases).

Hope that's acceptable.

**Comment by dergraf** [comment]
and how this is acceptable! :+1:

**Comment by mravkine** [comment]
Yes that would be great! Thanks guys.

**Comment by larshesel** [comment]
This just got merged to master and will be part of 1.4.2, unless we release 1.5.0 first (we may release 1.4.2 anyway).
